### PR TITLE
Fix subprocess handling

### LINKED ISSUE #47: Fix libypspur initialize timing
https://github.com/openspur/ypspur_ros/blob/6991c4c787aa8cce1ef992a1f0177d3fbff2aba5/src/ypspur_ros.cpp#L659-L664
this might fail if the system is busy

### LINKED ISSUE #49: Fix sub-process handling
causes dead lock in forked process sometimes on Alpine Linux (musl)

### PR DESCRIPTION
**Pull request by at-wat**
fix #47 
fix #49

### CODE REVIEW

File: src/ypspur_ros.cpp [hunk 1]
```diff
@@ -625,55 +628,78 @@ class YpspurRosNode
...
+        args.push_back(ad_mask);
+        args.push_back(std::string("--msq-key"));
+        args.push_back(std::to_string(key_));
+        if (digital_input_enable_)
+          args.push_back(std::string("--enable-get-digital-io"));
+        if (simulate_)
+          args.push_back(std::string("--without-device"));
+        if (param_file_.size() > 0)
         {
-          std::vector<std::string> args;
```
Comment (at-wat) [inline src/ypspur_ros.cpp]: Doing string processing after fork causes a kind of dead lock on strict environments like Alpine.

File: src/ypspur_ros.cpp [hunk 2]
```diff
@@ -1210,14 +1245,6 @@ class YpspurRosNode
       }
     }
     ROS_INFO("ypspur_ros main loop terminated");
-    ros::shutdown();
-    ros::spin();
-    if (pid_ > 0)
-    {
-      ROS_INFO("killing ypspur-coordinator (%d)", (int)pid_);
-      kill(pid_, SIGINT);
```
Comment (at-wat) [inline src/ypspur_ros.cpp]: without `wait`, child process goes zombie.

**Comment by at-wat** [comment]
@DaikiMaekawa please take a look

File: src/ypspur_ros.cpp [hunk 3]
```diff
@@ -625,55 +628,78 @@ class YpspurRosNode
...
+        msgctl(msq, IPC_RMID, nullptr);
 
-          execvp(ypspur_bin_.c_str(), const_cast<char **>(argv));
+        ROS_WARN("launching ypspur-coordinator");
+        pid_ = fork();
+        if (pid_ == -1)
+        {
+          const int err = errno;
+          ROS_ERROR("failed to fork process: %s", strerror(err));
+          throw(std::string("failed to fork process"));
```
Comment (DaikiMaekawa) [inline src/ypspur_ros.cpp]: Can you define a dedicated exception for these?
`std::string` could throw an exception during use.

File: src/ypspur_ros.cpp [hunk 4]
```diff
@@ -625,55 +628,78 @@ class YpspurRosNode
...
-          if (param_file_.size() > 0)
-          {
-            args.push_back(std::string("-p"));
-            args.push_back(param_file_);
-          }
+          args.push_back(std::string("-p"));
+          args.push_back(param_file_);
+        }
+
+        char **argv = new char *[args.size() + 1];
```
Comment (DaikiMaekawa) [inline src/ypspur_ros.cpp]: Where is it released?

File: src/ypspur_ros.cpp [hunk 5]
```diff
@@ -625,55 +628,78 @@ class YpspurRosNode
...
-        pid_ = fork();
-        if (pid_ == 0)
+        std::vector<std::string> args;
+        args.push_back(ypspur_bin_);
+        args.push_back(std::string("-d"));
+        args.push_back(port_);
+        args.push_back(std::string("--admask"));
+        args.push_back(ad_mask);
+        args.push_back(std::string("--msq-key"));
+        args.push_back(std::to_string(key_));
```
Comment (DaikiMaekawa) [inline src/ypspur_ros.cpp]: use the initializer list.

File: src/ypspur_ros.cpp (hunk 3 above)
Comment (at-wat) [reply src/ypspur_ros.cpp]: Many other part has same problem. I will fix them in separate PR. https://github.com/openspur/ypspur_ros/issues/51

Comment (DaikiMaekawa) [reply src/ypspur_ros.cpp]: fair enough.

**Comment by at-wat** [comment]
@DaikiMaekawa addressed your comments. PTAL

**Pull request merged**